lib/dry/validation/predicate_registry.rb

```diff
@@ -0,0 +1,82 @@
+module Dry
+  module Validation
+    class PredicateRegistry
+      attr_reader :predicates
+      attr_reader :external
+
+      class Bound < PredicateRegistry
+        def initialize(*args)
+          super
+          freeze
+        end
+      end
+
+      class Unbound < PredicateRegistry
+        def bind(schema)
+          bound_predicates = predicates.each_with_object({}) do |(n, p), res|
+            res[n] = Dry::Logic::Predicate.new(n, &p.bind(schema))
+          end
+          Bound.new(external, bound_predicates)
+        end
+
+        def update(other)
+          predicates.update(other)
+        end
+      end
+
+      def self.[](klass, predicates)
+        Unbound.new(predicates).tap do |registry|
+          klass.class_eval do
+            def self.method_added(name)
+              super
+              if name.to_s.end_with?('?')
+                registry.update(name => instance_method(name))
+              end
+            end
+          end
+        end
+      end
+
+      def initialize(external, predicates = {})
+        @external = external
+        @predicates = predicates
+      end
+
+      def [](name)
+        predicates.fetch(name) do
+          if external.key?(name)
+            external[name]
+          else
+            raise_unknown_predicate_error(name)
+          end
+        end
+      end
+
+      def key?(name)
+        predicates.key?(name) || external.key?(name)
+      end
+
+      def ensure_valid_predicate(name, args)
+        if key?(name)
+          predicate = self[name]
+
+          if predicate.arity != args.size + 1
```

Comment:
This will cause problems when we add support for predicates w/o args. We should probably check that arity !=0 before checking that there is space for the input arg.